### PR TITLE
x86: report when thread re-use is detected

### PR DESCRIPTION
**Pull request by tyrantlizard77**
x86_64's __resume path 'poisons' the incoming thread's
saved RIP value with a special 0xB9 value, to catch
re-use of thread objects across CPUs in SMP. Add a check
and printout for this when handling fatal errors.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Comment by andrewboie** [comment]
saw this come up once or twice...not sure the root cause, could be testcase problems, but worth making the poison scenario clearer

**Review by jhedberg — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

File: arch/x86/core/fatal.c [hunk 1]
```diff
@@ -155,6 +155,14 @@ static void dump_regs(const z_arch_esf_t *esf)
 #ifdef CONFIG_X86_EXCEPTION_STACK_TRACE
 	unwind_stack(esf->rbp, esf->cs);
 #endif
+#ifdef CONFIG_ASSERT
+	if (esf->rip == 0xb9) {
+		/* See implementation of __resume in locore.S. This is never
+		 * a valid RIP value.
+		 */
+		LOG_ERR("SMP: Re-use of thread object across CPUs found");
```
Comment (andyross) [inline arch/x86/core/fatal.c:163]: Personally I'd rephrase as "Attempt to run unsuspended thread object".  The core bug here isn't that the thread object got re-used necessarily, it's that something tried to switch to it when there was no context to switch to.  It could be a garbage pointer, an uninitialized thread, or (the original use case of the poisoning) a thread that is actively running on another CPU.

Comment (andrewboie) [reply arch/x86/core/fatal.c]: good idea

**Comment by andrewboie** [comment]
Updated per @andyross suggestion. Also always treat this as a kernel panic instead of some other kind of error.

**Review by andyross — APPROVED** [review]

**Review by nashif — APPROVED** [review]
